File: Python/Unique Binary Search Trees II.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @return a list of tree node
    def generate(self, left, right):
        if left > right:
            return [None]
        seq = []
        for i in range(left, right+1):
            lt = self.generate(left, i - 1)
            rt = self.generate(i + 1, right)
            for j in range(len(lt)):
                for k in range(len(rt)):
                    cnt = TreeNode(i)
                    cnt.left = lt[j]
                    cnt.right = rt[k]
                    seq.append(cnt)
        return seq
    def generateTrees(self, n):
        return self.generate(1, n)

File: C++/Unique-Binary-Search-Trees-II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
class Solution {
public:
    vector<TreeNode*> helper(int left, int right) {
        if (left > right) return vector<TreeNode*>(1, nullptr);
        vector<TreeNode*> vec;
        for (int k = left; k <= right; k ++) {
            vector<TreeNode*> lt = helper(left, k - 1);
            vector<TreeNode*> rt = helper(k + 1, right);
            for (int i = 0; i < lt.size(); i ++) {
                for (int j = 0; j < rt.size(); j ++) {
                    TreeNode* root = new TreeNode(k);
                    root->left = lt[i];
                    root->right = rt[j];
                    vec.push_back(root);
                }
            }
        }
        return vec;
    }
    vector<TreeNode*> generateTrees(int n) {
        return helper(1, n);
    }
};
